### DISCUSSION TITLE
Tabular MergeableStore persistence on Postgres/Sqlite

### DISCUSSION THREAD

**Discussion by guilhermeprokisch** [tinyplex/tinybase#189] [Ideas]
Hello there!

First off all thanks for this project; it has been incredibly helpful.

I was wondering if there are any plans in the future to support tabular MergeableStore persistence on Postgres/SQLite? I believe this would be amazing.

Currently, manipulating the state on the server side with the JSON CRDT format is quite challenging.

Or if there's at least a way to transform the CRDT json to a tabular form somehow.

**Comment by jamesgpearce** [comment]
So the challenge here is that even if we support a tabular view of the data in the database, there still needs to be a place to store the CRDT metadata (in the same columns or in parallel tables). Again, that will still be very difficult to modify correctly on the server unless it's done by TinyBase.

For example you wouldn't be able to just update a cell in a sqlite table and have it propagate because the the timestamp and hash in that other table would need to change (correctly) too.

The most fidelitous approach would be to have a TinyBase client on the server side that updates the data (and takes care of the metadata correctly). Open to other ideas here! (triggers? 🤣)

**Reply by guilhermeprokisch** [reply]
I'm not an expert in CRDTs or database performance, but is rly  wild to use triggers to implement the CRDT metadata, could works no?

Also, why does getMergeableContent return nested metadata? Couldn't we separate the CRDT metadata into its own table? That might make integrating with the existing persister easier. We could just adapt the load function to check if there's a metadata table. (Am I talking nonsense here?)

Another thought - could we expose triggers so people can add them on the Postgres or SQLite side directly in the database?

Does this make sense?
